### ISSUE TITLE
Researching modularization (gcloud-ruby perspective)

### ISSUE THREAD

**Issue by blowmage** [googleapis/google-cloud-common#138]
I spent some time this week diving into how to break `gcloud-ruby` apart into multiple packages, one package for each service. See also #30, #49, #52, GoogleCloudPlatform/gcloud-node#91, GoogleCloudPlatform/gcloud-python#1159, and GoogleCloudPlatform/gcloud-ruby#341. This would allow us to update code for one service's package without relesing a new package for the other services.

This means we would have a main `gcloud` gem, and it would have dependencies on the `gcloud-datastore`, `gcloud-storage`, `gcloud-pubsub`, and other gems. One of the (many) advantages of this approach is to be able to release pre-1.0 service gems that would not be included as a dependency in the main `gcloud` gem. For example, we could have had a `gcloud-search` gem that was not yet 1.0 while the service was still considered alpha.

One of the considerations for this approach is that all these packages must belong to the existing `gcloud-ruby` git repository. We will not be creating separate git repos for each package, instead we will be defining multiple packages in one shared repo. This is not a common approach for ruby projects, which means `gcloud-ruby` lose some tooling that it currently uses.

There are three main areas of change when moving to this approach: code changes, release process changes, and documentation changes. Here are some of the main issues I've identified for these areas. There may be more issues that I haven't thought of though.
## Code

Rails is one example of a repo that contains multiple gems (packages), so we will borrow from their approach. Each gem is a top-level directory in the repo. Early on individual gems were versioned and released as-needed, which is the approach we are wanting to take. It should be noted that starting in rails 2.0 they stopped individual versions and every main release (rails 4.2.3) has accompanying individual gem releases (activerecord 4.2.3, activesupport 4.2.3, etc.). This approach greatly simplifies the release process, and reduces the complications due to gems depending on each other.

Our approach differs in that each service gem can be released and versioned independently. This means that we will introduce a plugin mechanism to help with cross-package dependencies. This means individual gcloud gems will make themselves known to the main gcloud gem, so the UX will remain the same as the monolithic gem:

``` ruby
require "gcloud"
# gcloud-somenewservice is installed, but not required

gcloud = Gcloud.new
svc = gcloud.somenewservice # no require needed
```

There is some shared code that will be used by all service gems. The current thinking is to put this code in a gem named `gcloud-core` (`*-core` is the convention used by RSpec for it's shared code). Because service gems will be versioned and released independently different gems can have different requirements for the shared code. Therefore it is imperative this `gcloud-core` gem remains backwards compatible and will likely never reach 2.0.

All in all, I think this will be the easiest change to make.
## Release Process

Currently `gcloud-ruby` uses the Travis-CI release process to create the gem package and push to rubygems.org when a new tag is pushed to the git repo. When we break into separate service gems we will no longer be able to use their process, so we will have to write code to perform the releases.

We will also need to change the tag names, so that Travis-CI will know which package to build and push. Our current tags are `v0.10.0`, but my current thinking is that they will have to contain the package name/top-level directory as well as the version. E.g. `gcloud:1.2.0`, `gcloud-datastore:2.1.0`, `gcloud-storage:1.4.1`.

The way we use [GitHub releases](https://github.com/GoogleCloudPlatform/gcloud-ruby/releases) will also need to be updated, as the releases will be for an individual package (as identified by the git tag).
- Can't use default Travis code to release new gems, must write new code to do releases
- Can't use existing tags, must namespace tags so they include full gem name (gcloud-datastore-v1.2.0) so we know which gem the tag is a release of.

I think this will take a bit more time and effort than moving the existing code into multiple separate packages.
## Documentation

Maintaining our documentation is my largest area of conscern. The good news is that the JSON standard we've developed for types and methods should be able to be used with minimal changes. The bad news is that the angular application is going to become much more complicated.

Here are some of the issues I think need to be resolved:
- A way to resolve links to class/methods in different packages. E.g. Vision can annotate a Storage File, and we'd really like to be able to link from the Vision documents to the class in the Storage documents. Another common use-case is a method raising an error defined in the `gcloud-core` package.
- A way to specify version of service gems, as well as the main gem. If `gcloud` version 5.0 has a dependency on `gcloud-datastore` verison 2.0, it should link to that verson, even if `gcloud-datastore` verison 3.0 existed. This probably means we can't use a single version dropdown like we currently have.
- A way to list all service packages, even those not included in the main gem. So for pre-1.0 gems like the `gcloud-search` example earlier, the package was not included as a dependency in the main `gcloud` gem, but the documentation still needs to be discoverable. The documentation site should help users know what gems are part of the main package and what gems are not.

I think the documentation changes will take as much time and effort as the code and release process changes combined. This scares me.
## Final Thoughts

I think this approach has many advantages, but there are some pretty formidable trade-offs. If we were able to have purely separate git repos _and_ separate documentation sites then this would be much simpler. But maintaining the existing monolithic shell we've built up around the individual packages adds a tremendous amount of work. Despite my enthusiasm for modularizing the packages, I don't think its the best decision with the current requirements.

**Comment by quartzmo** [comment]
I also spent some time re-reading the commentary in the previous issues on this topic (listed below, please let me know if I missed any), and now this issue.

- GoogleCloudPlatform/gcloud-node#91- Modularizing
- GoogleCloudPlatform/gcloud-common#30- Should separate APIs be separately installable?
- GoogleCloudPlatform/gcloud-python#1159- Split out APIs into separately-installable distributions?
- GoogleCloudPlatform/gcloud-ruby#341- Should we publish a separate module (gem) for each Google Cloud service?
- GoogleCloudPlatform/gcloud-common#49 - Defining what 1.0.0 means
- GoogleCloudPlatform/gcloud-common#52- docs: individual module version management

In January [I commented](https://github.com/GoogleCloudPlatform/gcloud-common/issues/49#issuecomment-174736536):

> As hard as I try, I just can't imagine enough downside in practice to the single library approach. The issues that have been raised above seem quite thorny

I still feel this way. I think modularizing gcloud-ruby will add a burden of complexity, for both maintainers and users, for not much real benefit that I can see.

**Comment by jgeewax** [comment]
Keep in mind that the "killer feature" we're trying to provide is that people may way to pin to specific versions of specific packages, and use newer versions of newer packages. For example:

1. I write my code to talk to datastore (gcloud == 1.2.3)
1. I don't touch my code for a long time, it works as I want
1. Google releases magicservice and I want it, but I don't want to rewrite my code (and magicservice is in gcloud == 4.3.2)

At this point, I have two choices: either rewrite my existing stuff to use 4.x.x, or don't use magicservice. If we modularized the packages, this would be possible... Is there another way to accomplish this goal? (that isn't virtual environments... I don't want to manage a bunch of env's just to use magicservice)

**Comment by jgeewax** [comment]
### Meta-package (ie, `gcloud`) versions

As previous discussed, we should publish common components in a "core" package, while not useful on it's own, supplies shared common functionality like base API clients, JSON parsing, error handling, iterators, etc.

This might look like....

```
$ npm install gcloud-core
```

And then publish each API as it's own package in the package manager
```
$ npm install gcloud-datastore
```

And finally, publish an "umbrella" metapackage containing all service packages
```
$ npm install gcloud
```

But this leads to a few concerns about versioning. As we've said before, following semver should work. In other words, we should base our version changes on the following rule:

> Will upgrading to this version of the umbrella metapackage break customer code? If so, increment the major version.

Increasing the major version of the meta-package means that there may be scenarios where the meta-package major version surpasses the major versions of the service packages, potentially by a lot (ie, the worst case scenario is that `version(gcloud) = sum([version(gcloud-<service>) for all services])`)

#### What about auto-generated code?

As we start to use more of the auto-generated code, we have to consider those versions as well. Luckily, the same principles apply to the auto-generated package, however the package versioning only applies to the version of the generator and not the version of the service. As new service versions are launched (and we generate new client libraries based on the same generator), the new code could be put into a subdirectory named after the service version (ie, `v2/`), resulting in a minor version increase due to a purely additive change (ie, `1.0.8` -> `1.1.0`).

For example, if the auto-generated PubSub package is at version `1.0.0` and contains just PubSub's `v1` API surface, and PubSub launches a `v2` API, the auto-generated package would generate code for `v2`, and put it in auto-generated package such that it's import path would be from `gcloud.pubsub.v2`. By doing that, the auto-generated PubSub package would increase it's minor version to `1.1.0` since the change is backwards compatible.

#### What about `gcloud-core` version conflicts?

Since the `gcloud-core` package would be shared across all of the service packages, it's possible to get into a state where the customer wants versions of two different service packages that require different major versions of the core package.

One simple way to address this is to guarantee that we'll never get past major version 1 (in other words, that we'll always be backwards compatible). I don't know if that will always work, so we need a plan for when we're dealing with backwards incompatible changes.

If we happen to be forced to make a change that is backwards incompatible, and since we expect the core package to be shared across multiple different versions of service packages, it'd probably be a good idea to namespace major versions in the package akin to how the auto-generated versioning works.

For example, usage of the `gcloud-core` library in Python would look like:

```
$ pip install gcloud-core~="2.0"
>>> from gcloud.core.v1 import BaseClient
$ pip install "gcloud-core~=3.0"
>>> from gcloud.core.v1 import BaseClient
>>> from gcloud.core.v2 import BaseClient
```

#### An example....

|What happened? | `gcloud` | `gcloud-pubsub` | `gcloud-magicservice`
---|---|---|---
- | 1.5.0 | 1.3.1 | -
Bug fix to PubSub client | 1.5.1 | 1.3.2 | -
Created first pass at magicservice | 1.5.1 | **1.3.2** | 0.1.0
PubSub uses API v5 | **1.6.0** | **1.4.0** | 0.1.0
Stable RC of magicservice client | 1.6.0 | 1.4.0 | **1.0.0-beta1**
PubSub using v6, with breaking changes | **2.0.0** | **2.0.0** | 1.0.0-beta1
magicservice is now stable | 2.1.0 | 2.0.0 | **1.0.0**
We auto-gen'd code for PubSub v6 | **2.2.0** | **2.1.0** | 1.0.0
Launched a new generator (backwards incompatible) | **3.0.0** | **3.0.0** | 1.0.0

**Comment by quartzmo** [comment]
@jgeewax Regarding the motivation for this feature:

> pin to specific versions of specific packages, and use newer versions of newer packages

Is it not likely that Cloud Platform APIs will rapidly stabilize and then exhibit very low volatility? (The service versions in [aws-sdk-ruby](https://github.com/aws/aws-sdk-ruby#supported-services) suggest such stability.) This would imply that, in the example given, migrating from gcloud `1.2.3` to `4.3.2` would very likely **not** require any change to existing datastore code? Furthermore, in cases where changes to client code and service version are required, it would not best serve the business to motivate customers to make these upgrades?

(As an engineer, I am enthusiastic about this feature--it is cool and sophisticated. However, I feel obliged to play devil's advocate regarding its benefits. Please consider these questions just routine due diligence. I do not presume to know the priorities of the business.)

**Comment by quartzmo** [comment]
@callmehiphop @stephenplusplus You've been considering this [for a long while](https://github.com/GoogleCloudPlatform/gcloud-node/issues/91), any comments? Anything to add?

**Comment by stephenplusplus** [comment]
Here are some thoughts from a Node perspective:

>  @jgeewax: As we've said before, following semver should work.

👍 

> @jgeewax: What about gcloud-core version conflicts?

npm flattens dependencies as far as it can, but if we have end up needing two different versions of gcloud-core, npm will give them to us.

-- 

I modified this table to show that if we use the right semver ranges, we don't have to bump `gcloud` as much.

|What happened? | `gcloud` | `gcloud-pubsub` | `gcloud-magicservice`
---|---|---|---
- | 1.5.0 | 1.3.1 | -
Bug fix to PubSub client | ~~1.5.1~~ 1.5.0 new installs will automatically pick up the `gcloud-pubsub` fix without requiring another `gcloud` release | **1.3.2** | -
Created first pass at magicservice | ~~1.5.1~~ **1.6.0** this should be a feature | 1.3.2 | 0.1.0
PubSub uses API v5 | ~~**1.6.0**~~ 1.6.0 This doesn't require any bump. `gcloud` depends on `gcloud-pubsub@^1.0.0` which means "> 1.0.0 && < 2.0.0". So, new installs would have picked this up automatically as well. | **1.4.0** | 0.1.0
Stable RC of magicservice client | 1.6.0 | 1.4.0 | **1.0.0-beta1**
PubSub using v6, with breaking changes | **2.0.0** | **2.0.0** | 1.0.0-beta1
magicservice is now stable | **2.1.0** Just a note that if it being stable means it introduced any breaking changes from 0.1.0, this would have to bump to **3.0.0** | 2.0.0 | **1.0.0**
We auto-gen'd code for PubSub v6 | ~~**2.2.0**~~ 2.1.0 - No bump necessary because we will be depending on `gcloud-pubsub@^2.0.0`, which picks up new minor versions automatically. | **2.1.0** | 1.0.0
Launched a new generator (backwards incompatible) | **3.0.0** | **3.0.0** | 1.0.0

--

If I understood #139 correctly, we are all using our one repo to hold all of our modules. So, the action items for Node:

- Create a package.json for each service that:
  - includes only the "common" files and the service files
  - has a pre-publish script to:
    - build doc json & send to gh-pages
    - if it was a breaking change...
      - bump `gcloud`'s dependency on it
      - publish `gcloud`
  - publish each service

Doc design changes to accommodate modularization have been proposed here: #52. Each service will have its own page, so as to avoid confusion between:

gcloud-node -> Storage (`gcloud-storage@0.2.0`) -> File
... and ...
gcloud-storage (`gcloud-storage@0.1.1`) -> File

When the pre-publish script builds the docs for a new `gcloud-storage` release, it will add a new version to its own manifest.json. The docs for the `gcloud-node` (parent package) should use its own `package.json` to understand what folder inside of `gh-pages/storage/json/` to load JSON from. (This effectively means when you load `gcloud-node` (parent package)'s docs, it will look for the most accurate JSON to load from each service.

@callmehiphop -- thoughts on all of this please!

**Comment by callmehiphop** [comment]
LGTM - my only real suggestion would be to avoid making multiple manifest files, instead move versions to a <kbd>versions.json</kbd> file (or similar) to eliminate configuration duplications?
